modules/table/src/main/java/org/apache/ignite/internal/table/distributed/replicator/IndexBuilderTxRwOperationTracker.java

```diff
@@ -0,0 +1,203 @@
+
+            futures.clear();
+        });
+    }
+
+    /**
+     * Waits for RW transactions operations to complete before the requested catalog version.
+     *
+     * <p>It first waits for the requested catalog version to be reached in {@link #updateMinAllowedCatalogVersionForStartOperation(int)}
+     * and then waits for all RW transaction operations to complete.</p>
+     *
+     * @param catalogVersion Catalog version in which the new index appeared.
+     */
+    CompletableFuture<Void> awaitCompleteTxRwOperations(int catalogVersion) {
+        return inBusyLock(busyLock, () -> {
+            // This code is needed to avoid races with updateAllowedCatalogVersionForStartOperation.
+            CompletableFuture<Void> awaitFuture = awaitCompleteOperationsFutureByCatalogVersion.computeIfAbsent(
+                    catalogVersion,
+                    i -> new CompletableFuture<>()
+            );
+
+            if (catalogVersion <= minAllowedCatalogVersionForStartOperation.get()) {
+                awaitCompleteOperationsFutureByCatalogVersion.remove(catalogVersion);
+
+                awaitFuture.complete(null);
+            }
+
+            // Let's prevent freezing.
+            awaitFuture = orCloseFuture(awaitFuture);
+
+            return awaitFuture.thenCompose(unused -> {
+                // Let's select counters for versions up to (not including) the requested catalog version.
+                CompletableFuture<?>[] futures = operationCounterByCatalogVersion.headMap(catalogVersion, false).values().stream()
+                        .map(TxRwOperationCounter::operationsFuture)
+                        .toArray(CompletableFuture[]::new);
+
+                // Let's prevent freezing.
+                return orCloseFuture(allOf(futures));
+            });
+        });
+    }
+
+    /**
+     * Increases the count of transaction operations by {@code 1}, at the start of a new RW transaction operation.
+     *
+     * <p>It is expected that after the operation is completed, {@link #decrementOperationCount(int)} will be called.</p>
+     *
+     * @param catalogVersion Catalog version at the beginning of the RW transaction of operation.
+     * @throws RejectTxRwOperationException If the passed catalog version is less than the minimum available to start a RW transaction
+     *      operation.
+     */
+    void incrementOperationCount(int catalogVersion) {
+        inBusyLock(busyLock, () -> {
+            operationCounterByCatalogVersion.compute(catalogVersion, (i, txRwOperationCounter) -> {
+                if (txRwOperationCounter == null) {
+                    return new TxRwOperationCounter();
+                }
+
+                return txRwOperationCounter.incrementOperationCount();
+            });
+
+            if (catalogVersion < minAllowedCatalogVersionForStartOperation.get()) {
+                decrementOperationCount(catalogVersion);
+
+                throw new RejectTxRwOperationException();
```

Comment:
At the moment, I assume that this exception will be thrown rarely and if it is thrown, the runtime costs will not be greatly affected.

Reply:
But returning a boolean is not more difficult and it's more natural. What are the benefits of using an exception? It breaks the 'don't use exceptions for control logic' rule.

Reply:
At the moment I can’t agree with you that your proposal is more natural, for example methods `AtomicInteger#getAndIncrement` and `AtomicInteger#incrementAndGet` returns integere and not a logical value. 
If in practice it turns out that your approach will be better and more optimal, then we will change it.

Reply:
Let's compare apples to apples. The methods you mentioned are not part of the CAS API. They cannot 'fail' their update due to a concurrent update changing something under their nose.

`incrementOperationCount()` is something different: it's a clear CAS method as it can either succeed or fail due to concurrency; also it doesn't return anything on its own, so we can just add the `boolean` return type without any efforts. `compareAndSet()` is an example of a CAS method, and it returns exactly `boolean` to communicate success or failure. This is why I suggest to return `boolean` here as well.

Reply:
Removed the exception and replaced it with a boolean value.